README

amfext (native) 0.8.*
=====================

This extension implements the ActionScript Message Format 
for encoding and decoding,both in AMF and AMF3 versions.
Information can be found in [0].

This extension provides two low level functions useful for
encoding and decoding objects into AMF format, but it does
not provide functions for managing the full message. An
example of implementation can be found in [1].

Details on the encoding process can be read in doc/encoding.txt

Documentation on the PHP function and the callbacks is in doc/amfextdoc.php

Constants for the flags that can be used in the encoding/decoding are in amfext.php
that can be included.

The directory tests contains some tests that can be invoked directly. Eventually drop AMFPHP
sources somewhere for testing the PHP version against the C version. See test_common.php for
details.

Building
-------------

Under Windows: download the source of your PHP version and unpack the zip under the
ext directory of the source. Open the associated Visual Studio project (amf4.dsp or amf5.dsp)
and just build. Everything should be fine. For distributing the extension is suggested to use 
Visual Studio 6, otherwise 2003 is fine.

Under Uni*/Linux: obtain your php-devel package and unpack the zip in a directory. Then
as any PHP extension perform:

    phpize
    ./configure --enable-amf
    make install
    
Note: there could be some little problems if php includes are not under /usr

In any case remember to add the amf.so/php_amf.dll extension to the php.ini file

Known Builds
--------------

This code has been tested under Windows (PHP 4.4.0, 5.2.0) and Fedora (PHP 4.1.6) under a 32 bit machine.

Please check the behaviour with 64-bit machine

Thanks
---------------

Thanks to Patrick Mineault from AMF for the support, and from the authors
of the JSON PHP extension for the skeleton of the project.


Authors
-------

Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>

---

[0] http://osflash.org/doku.php?id=amf
[1] http://amfphp.sourceforge.net/


ChangeLog
2008-07-17  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version 0.9.1
- fixed pecl bug 12668 on anonymous serialization

2007-02-21  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version 0.9
- added "version" in "serverInfo" for AM0 while sending recordset
- charset handling

2007-02-20  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- cleaned comments, AMF_DEALLOC macro and if

2007-02-18  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- added "version" in "serverInfo" for AM0 while sending recordset

2007-02-16  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version 0.9
- changed back the callback
- fixed reference issue in callback, caused by callback function with &

2007-02-14  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- finished implementation of the charset handling
- small test with testCharset
- fixed all tests with the new serialization callback
- added AMF_TRANSLATE_CHARSET_NOASCII in decoding to skip UTF8 that are US-ASCII (that is lower than 0x7F): 
	UTF-8, ISO-8859-1 have the first part in common to US-ASCII
	
	AMF_TRANSLATE_CHARSET_NOASCII cannot be used in encoding Big5
		Big5 is double byte: first byte is ASCII or in 0x81-0xFE, second in 0x40 to 0x7e, 0xa1 to 0xfe
		Shift_JIS is double byte: first byte is is ASCII (with exception) or over, second can be low or high	
  

2007-02-13  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- initial implementation of the charset handling. 
- TODO: check the use of shared
- TODO: check the new scheme for encoding callback
- TODO: check the use of raw

2007-02-08  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version 0.8.9
- added AMF_TRANSLATE_CHARSET flag, AMFE_TRANSLATE_CHARSET event and added event to the callback
for supporting different charsets. Changes have been applied to: amf3_write_string,amf3_write_string_zval
amf0_serialize_string,amf0_serialize_zstring,amf0_write_string. In AMF3 the transliteration is performed
AFTER the cache check. TODOOOOO!

2007-02-01  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version 0.8.8
- StringBuilder as resource
- AMF_AS_STRING_BUILDER in amf_encode for returing a StringBuilder resource
- fixed bug in the amf4.dsp creating bad PHP4 modules

2007-01-27  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- removed warning in linux build

2007-01-20  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version 0.8.7
- fixed encoding of private member variables 

2007-01-13  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>

- fixed head of AMF3_OBJECT that was written as byte
- fixed case of rows shorter than columns in recordset

2007-01-12  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version  0.8.5 and change name to AMF
- fixed IS_ARRAY

2007-01-11  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
version: 0.8.3
- added AMFC_BYTEARRAY in encoding callback for encoding ByteArray
- added AMFE_BYTEARRAY for decoding ByteArray as a callback
- no more AMFE_MAP for objects with empty classname

2007-01-08  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- added phpinfo
- added nextStringIndex for future optimized zval lookup
- 
- fixed caching of zval with handling of is_ref

2007-01-07  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- fixed return of classname from encoding callback
- added __amfphp_recordset__ for AMF0 exposing it as RecordSet object
- fixed array count in decoding array

2007-01-06  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>

- arrays inited to 
- stdclass class is mapped to anonymous object in AMF3
- added AMFE_POST_XML
- added AMFE_MAP_EXTERNABLIZABLE
- negative integer keys 

2007-01-05  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>

- added support for resources
- fixed type check in encoding callbacks
- added documentation
- cleaned numeric string array test
    
2007-01-02  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>
- optimized encoding by replacing smart_str with chunked string builder
- fixed arrays in AMF3
- corrected reference counts

2006-12-27  Emanuele Ruffaldi <emanuele.ruffaldi@gmail.com>

- Initial version


config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Whether to build amf as dynamic module */
#define COMPILE_DL_AMF 1

/* whether to have AMF Object Serialization support */
#define HAVE_AMF 1

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to the sub-directory in which libtool stores uninstalled libraries.
   */
#define LT_OBJDIR ".libs/"

/* Define to 1 if your C compiler doesn't accept -c and -o together. */
/* #undef NO_MINUS_C_MINUS_O */

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1


amf_sb.php
<?php

// it provides the functions for the AMF StringBuilder based on strings
if(!extension_loaded("amf"))
{
	function amf_sb_new()
	{
		return "";
	}

	function _amf_sb_append(&$sb,$a)
	{
		foreach($a as $v)
		{
			if(is_array($v))
				_amf_sb_append($sb,$v);
			else
				$sb .= $v;
		}
	}
	
	function amf_sb_append(&$sb)
	{
		$n = func_num_args();
		$r = $sb;
		for($i = 1; $i < $n; $i++)
		{
			$aa = func_get_arg($i);
			if(is_array($aa))
			{
				_amf_sb_append($r,$aa);
			}
			else
				$r .= $aa;
		}
		$sb = $r;
	}

	function amf_sb_length(&$sb)
	{
		return strlen($sb);
	}
	
	function amf_sb_write(&$sb,$stream=NULL)
	{
		if($stream == NULL)
			echo($sb);
		else
			fwrite($stream,$sb);
	}
    
    function amf_sb_echo(&$sb)
	{
		echo($sb);
	}
	
	function amf_sb_as_string(&$sb)
	{
		return $sb;
	}
    
	function amf_sb_flat(&$sb)
	{
		return $sb;
	}
}

?>

amfext.php
<?php
/**
 * amfphp encoding and decoding of AMF and AMF3 data
 * 
 * @license http://opensource.org/licenses/php.php PHP License Version 3
 * @copyright (c) 2006-2007 Emanuele Ruffaldi emanuele.ruffaldi@gmail.com
 * @author Emanuele Ruffaldi
 *
 * This is an import file for constants defined by AMFPHP C
 */

/// AMFPHP native flags for encode and decode
define("AMF_AMF3",1);   // encoding: use AMF3,   decoding: AMF3 was found
define("AMF_BIG_ENDIAN",2);  // encoding/decoding: machine is bigendian
define("AMF_ASSOCIATIVE_DECODE",4); // decoding: treat anonymous objects as associative arrays
define("AMF_POST_DECODE",8); // decoding: invoke post decoding callback on every object handled as object
define("AMF_AS_STRING_BUILDER",16); // decoding: invoke post decoding callback on every object handled as object
define("AMF_TRANSLATE_CHARSET",32); // encoding/decoding: translate every string
define("AMF_TRANSLATE_CHARSET_NOASCII",32|64); // skips US-ASCII strings for translation, only in decoding

/// AMFPHP native event types for decoding callback
/// The decoding callback has two parameters. The first is the type of event from the following and the others is the argument of the event
///    
/// AMFE_MAP event: invoked when a typed object is received it allows to map an AMF className into a PHP classname or array. The argument is the name of the class and the callback should return the resulting
///  container as an object or array. The callback is not invoked if the object has an empty classname
/// pseudocode:
///         $r = mycallback(AMFE_MAP, $classname)
///         if(not $r is ARRAY and not $r is OBJECT)
///             $r = new $classname()
/// AMFE_POST_OBJECT event: invoked after the decoding of an Object only if the AMF_POST_DECODE flag has been set. The argument is the value of the object that can be modified by the callback
///          $r = mycallback(AMFE_POST_DECODE, $r)
/// AMFE_POST_XML event: invoked after the decoding of an XML structure. The argument is a string
/// AMFE_MAP_EXTERNALIZABLE: invoked for mapping a class that is externalizable. $arg is the classname and the callback should return the resulting
///  container as an object or array
/// AMFE_POST_BYTEARRAY: invoked for mapping a ByteArray object, otherwise it is represented as a string
/// AMFE_TRANSLATE_CHARSET: invoked for mapping a string into another string using charset
define("AMFE_MAP",1);
define("AMFE_POST_OBJECT",2);
define("AMFE_POST_XML",3);
define("AMFE_MAP_EXTERNALIZABLE",4);
define("AMFE_POST_BYTEARRAY",5);
define("AMFE_TRANSLATE_CHARSET",6);

define("AMFR_NONE",0); 
define("AMFR_ARRAY",1); 
define("AMFR_ARRAY_COLLECTION",2); 

/// AMFPHP callback return types 
/// The encoding callback has two parameters and returns a single value or an array with up to 3 elements:
///      (value, type, classname) = mycallback(value, classname)
/// Value is the new value to be encoded
/// Type is the type of the value from the following types, if it is not specified it is AMFC_TYPEDOBJECT
/// Classname is the name of the class to be used
///
define("AMFC_RAW", 0); // treat the returned string as raw AMF data
define("AMFC_XML", 1); // treat the returned string as XML data
define("AMFC_OBJECT",2); // treat the returned value as an anonymous object
define("AMFC_TYPEDOBJECT",3); // treat the returned value as an object with type. The class is specified by the third parameter
define("AMFC_ANY",4); // interpret again the returned value
define("AMFC_ARRAY",5); // treat the returned value as an array
define("AMFC_NONE",6); // returns undefined
define("AMFC_BYTEARRAY",7); // returns a ByteArray but only in AMF3

?>